docs/_tooling/extensions/header_service/header_service.py

```diff
@@ -0,0 +1,309 @@
+# *******************************************************************************
+# Copyright (c) 2025 Contributors to the Eclipse Foundation
+#
+# See the NOTICE file(s) distributed with this work for additional
+# information regarding copyright ownership.
+#
+# This program and the accompanying materials are made available under the
+# terms of the Apache License Version 2.0 which is available at
+# https://www.apache.org/licenses/LICENSE-2.0
+#
+# SPDX-License-Identifier: Apache-2.0
+# *******************************************************************************
+from collections import defaultdict
+import os
+import re
+import subprocess
+import hashlib
+import random
+from typing import Any
+
+from github import (
+    Auth,
+    Github,
+    Organization,
+    PaginatedList,
+    PullRequestReview,
+    Repository,
+)
+from sphinx_needs.services.base import BaseService
+from sphinx_needs.data import SphinxNeedsData
+from sphinx.application import Sphinx
+from sphinx.util.docutils import SphinxDirective
+
+# req-traceability: GD__automatic_document_header_generation
+
+APPROVER_TEAMS = ["Steering Council", "WG Process"]
+
+
+def setup(app: Sphinx) -> dict[str, str | bool]:
+    """Register the header service with the Sphinx application.
+
+    :param app: The Sphinx application instance.
+    """
+    app.connect("env-before-read-docs", register)
+    return {
+        "version": "0.1",
+        "parallel_read_safe": True,
+        "parallel_write_safe": True,
+    }
+
+
+def register(app: Sphinx, env: SphinxNeedsData, a: str) -> None:
+    """Register the HeaderService with the Sphinx application environment.
+
+    :param app: The Sphinx application instance.
+    :param env: The Sphinx environment.
+    :param a: Additional arguments.
+    """
+    app.add_config_value("header_service_use_github_data", True, "env")
```

Comment:
This will be executed for every single "env-before-read-docs". Is this really intentional or can we register the service only once?

Reply:
Ok, so it's not per file. That's already good :-)
But keep in mind that 1-17 is basically an endless loop. Whatever we can move to setup() we should.